### DISCUSSION TITLE
Help! dev/prod request headers inconsistencies

### DISCUSSION THREAD

**Discussion by bpolaszek** [nuxt/nuxt#26084] [Questions] [answered]
Hello there !

I am about to release my first Nuxt project in production, but I noticed some inconsistencies between _development_ and _production_ mode.
My app has a conditional rendering based on the _hostname_ it's been called from. 
I'm using `useRequestURL().hostname` for that, and it works well both server-side and client-side... in `development` mode only.

In `production` mode (`yarn build && yarn preview`), `useRequestURL().hostname` always resolves to `localhost`. 😨 

I also tried to console.log `useRequestHeaders(['x-forwarded-host', 'host'])`:
I get the appropriate values when I `yarn dev` but I get an empty object in production.

This only happens on the server-side, client side is OK.

Any ideas?

**Comment by xjccc** [comment]
nginx set header?
```
{
    proxy_set_header Host $host;
}

```

**Reply by bpolaszek** [reply, marked as answer]
The problem doesn't come from the web server, since I get the appropriate headers (and  `useRequestURL().hostname`) with `nuxt dev`.  It just doesn't work when I use `nuxt preview`.

I figured out a solution by adding a nitro middleware which will hydrate a context property, but I consider we shouldn't require this, since `useRequestURL()` has been made for this purpose, so I think it deserves filing an issue.

For whoever stumbles on the same problem:

```ts
// server/middleware/hostname.ts
export default defineEventHandler(({req, res, context}) => {
  context.hostname = req.headers.host
})
```

Usage:
```ts
// composables/useHostname.ts
export const useHostname = () => {
  return useNuxtApp().ssrContext?.event?.context?.hostname // When SSR
    ?? useRequestURL().hostname // When client
}
```

**Reply by bpolaszek** [reply]
Additional context: looks like this bug appears using route rules, regardless of `development` / `production`. I actually had a `swr` rule which was conditionally `0` or `3600` depending on the environment.
I just filed a bug report about that - https://github.com/nuxt/nuxt/issues/26102
